scheduler/src/interfaces/server/controller_interface.rs

```diff
@@ -0,0 +1,126 @@
+        }
+    }
+}
+
+type ScheduleActionStream = ReceiverStream<Result<proto::ActionResponse, tonic::Status>>;
+
+#[tonic::async_trait]
+impl Controller for ControllerService {
+    type ScheduleActionStream = ScheduleActionStream;
+
+    async fn schedule_action(
+        &self,
+        request: tonic::Request<proto::ActionRequest>,
+    ) -> Result<tonic::Response<Self::ScheduleActionStream>, tonic::Status> {
+        let action_request = request.into_inner();
+
+        let context = match action_request.context.clone() {
+            Some(context) => context,
+            None => {
+                warn!("Context field is missing for ActionRequest {}", action_request.action_id);
+                return Err(tonic::Status::invalid_argument("Context field is missing"));
+            }
+        };
+
+        let runner_type = match context.r#type.into() {
+            Some(runner_type) => runner_type,
+            None => {
+                warn!("Invalid RunnerType in ExecutionContext for ActionRequest {}", action_request.action_id);
+                return Err(tonic::Status::invalid_argument("Invalid RunnerType"));
+            }
+        };
+        
+        let container_image = match context.container_image.clone() {
+            Some(container_image) => container_image,
+            None => {
+                warn!("ContainerImage field is missing for ActionRequest {}", action_request.action_id);
+                return Err(tonic::Status::invalid_argument("ContainerImage field is missing"));
+            }
+        };
+
+        info!("Received Context action request: {},\n
+            Context runner type: {}", context.container_image.unwrap(), runner_type);
+
+        // Lock the Agent Pool (to ensure thread-safe access). This is a tokio Mutex, not a standard one.
+        let pool = self.agent_pool.lock().await;
+
+        // Same for the Action Queue
+        let mut queue = self.action_queue.lock().await;
+
+        // Create a new Action and add it to the queue (it gets sorted)
+        let new_action = Action::new(
+            action_request.action_id,
+            proto::ExecutionContext {
+                container_image: Some(container_image),
+                r#type: runner_type,
+            },
+            action_request.commands,
+        );
+
+        // Add the Action to the Action Queue
+        queue.push(new_action);
+
+        // Get the Agent with the lowest score from the Agent Pool
+        let agent = match pool.peek() {
+            Some(agent) => agent,
+            None => {
+                warn!("No agents available to execute action");
+                return Err(tonic::Status::unavailable("No agents available"));
+            }
+        };
+
+        // Loop over the action queue
+        while let Some(action) = queue.pop() {
+            // Send the action to the Agent using agent_client.rs
+            if let Err(e) = agent_client::execution_action(action, agent.get_ip_address()).await {
```

Comment:
You should not put your agent choice logic here as you need to give logs of the scheduled action back to the controller.

Reply:
Any news ?

Reply:
it is marked as outdated and i don't find the related code i think that you should review the actual code to specify your comment on the updated code